File: backend/shared/schemas/device.py
"""Pydantic schemas for Device."""

from pydantic import BaseModel, Field, field_validator, IPvAnyAddress
from typing import Optional
from datetime import datetime
from enum import Enum


class DeviceStatus(str, Enum):
    """Device status enumeration for Pydantic validation."""

    ONLINE = "online"
    OFFLINE = "offline"
    UNKNOWN = "unknown"


class DeviceBase(BaseModel):
    """Base schema with common device fields."""

    name: str = Field(..., min_length=1, max_length=200, description="Device display name")
    ip_address: str = Field(..., description="Device IP address (IPv4 or IPv6)")
    port: int = Field(default=4370, ge=1, le=65535, description="Device port (default: 4370)")
    serial_number: Optional[str] = Field(None, max_length=100, description="Device serial number")
    location: Optional[str] = Field(None, max_length=200, description="Device location")
    description: Optional[str] = Field(None, description="Device description")

    @field_validator("ip_address")
    @classmethod
    def validate_ip_address(cls, v: str) -> str:
        """Validate IP address format."""
        try:
            IPvAnyAddress(v)
            return v
        except Exception:
            raise ValueError("Invalid IP address format")


class DeviceCreate(DeviceBase):
    """Schema for creating a new device."""

    device_group_id: Optional[int] = Field(None, description="Optional device group ID")


class DeviceUpdate(BaseModel):
    """Schema for updating device (all fields optional)."""

    name: Optional[str] = Field(None, min_length=1, max_length=200)
    ip_address: Optional[str] = None
    port: Optional[int] = Field(None, ge=1, le=65535)
    serial_number: Optional[str] = Field(None, max_length=100)
    location: Optional[str] = Field(None, max_length=200)
    description: Optional[str] = None
    device_group_id: Optional[int] = None

    @field_validator("ip_address")
    @classmethod
    def validate_ip_address(cls, v: Optional[str]) -> Optional[str]:
        """Validate IP address format if provided."""
        if v is None:
            return v
        try:
            IPvAnyAddress(v)
            return v
        except Exception:
            raise ValueError("Invalid IP address format")


class DeviceResponse(DeviceBase):
    """Schema for device response."""

    id: int
    school_id: int
    status: DeviceStatus
    last_seen: Optional[datetime] = None
    last_sync: Optional[datetime] = None
    max_users: Optional[int] = None
    enrolled_users: int = 0
    device_group_id: Optional[int] = None
    created_at: datetime
    updated_at: Optional[datetime] = None

    class Config:
        from_attributes = True


class DeviceListResponse(BaseModel):
    """Paginated list of devices."""

    items: list[DeviceResponse]
    total: int
    page: int
    page_size: int
    pages: int


class DeviceConnectionTest(BaseModel):
    """Schema for connection test request."""

    timeout: Optional[int] = Field(
        default=5, ge=1, le=30, description="Connection timeout in seconds"
    )


class DeviceConnectionTestResponse(BaseModel):
    """Schema for connection test response."""

    success: bool
    message: str
    device_info: Optional[dict] = None  # Device information if connection successful
    response_time_ms: Optional[int] = None

File: frontend/lib/validations/device.ts
/**
 * Zod validation schemas for Device entities.
 * 
 * Matches the backend Pydantic schemas in shared/schemas/device.py
 */

import { z } from 'zod';

/**
 * Device status enumeration.
 */
export const deviceStatusEnum = z.enum(['online', 'offline', 'unknown']);

/**
 * IP address validation (IPv4 or IPv6).
 */
const ipAddressSchema = z
  .string()
  .min(1, 'IP address is required')
  .refine(
    (val) => {
      // IPv4 regex
      const ipv4Regex = /^((25[0-5]|2[0-4][0-9]|[01]?[0-9][0-9]?)\.){3}(25[0-5]|2[0-4][0-9]|[01]?[0-9][0-9]?)$/;
      // IPv6 regex (simplified - matches standard IPv6 formats)
      const ipv6Regex = /^([0-9a-fA-F]{1,4}:){7}([0-9a-fA-F]{1,4}|:)$|^::1$|^::$|^([0-9a-fA-F]{1,4}:)*::([0-9a-fA-F]{1,4}:)*([0-9a-fA-F]{1,4})?$|^([0-9a-fA-F]{1,4}:)*::([0-9a-fA-F]{1,4}:)*([0-9a-fA-F]{1,4})?$/;
      
      return ipv4Regex.test(val) || ipv6Regex.test(val);
    },
    { message: 'Invalid IP address format (must be IPv4 or IPv6)' }
  );

/**
 * Device form validation schema (for create).
 */
export const deviceFormSchema = z.object({
  name: z
    .string()
    .min(1, 'Device name is required')
    .max(200, 'Device name must be less than 200 characters'),
  
  ip_address: ipAddressSchema,
  
  port: z
    .number()
    .int('Port must be an integer')
    .min(1, 'Port must be between 1 and 65535')
    .max(65535, 'Port must be between 1 and 65535')
    .default(4370),
  
  com_password: z
    .string()
    .max(20, 'Communication password must be less than 20 characters')
    .nullable()
    .optional()
    .transform((val) => val === '' ? null : val),
  
  serial_number: z
    .string()
    .max(100, 'Serial number must be less than 100 characters')
    .nullable()
    .optional()
    .transform((val) => val === '' ? null : val),
  
  location: z
    .string()
    .max(200, 'Location must be less than 200 characters')
    .nullable()
    .optional()
    .transform((val) => val === '' ? null : val),
  
  description: z
    .string()
    .nullable()
    .optional()
    .transform((val) => val === '' ? null : val),
  
  device_group_id: z
    .number()
    .int()
    .positive()
    .nullable()
    .optional(),
});

/**
 * Device create form data type.
 */
export type DeviceFormData = z.infer<typeof deviceFormSchema>;

/**
 * Device update form data (all fields optional).
 */
export const deviceUpdateSchema = z.object({
  name: z
    .string()
    .min(1, 'Device name is required')
    .max(200, 'Device name must be less than 200 characters')
    .optional(),
  
  ip_address: ipAddressSchema.optional(),
  
  port: z
    .number()
    .int('Port must be an integer')
    .min(1, 'Port must be between 1 and 65535')
    .max(65535, 'Port must be between 1 and 65535')
    .optional(),
  
  com_password: z
    .string()
    .max(20, 'Communication password must be less than 20 characters')
    .nullable()
    .optional()
    .transform((val) => val === '' ? null : val),
  
  serial_number: z
    .string()
    .max(100, 'Serial number must be less than 100 characters')
    .nullable()
    .optional()
    .transform((val) => val === '' ? null : val),
  
  location: z
    .string()
    .max(200, 'Location must be less than 200 characters')
    .nullable()
    .optional()
    .transform((val) => val === '' ? null : val),
  
  description: z
    .string()
    .nullable()
    .optional()
    .transform((val) => val === '' ? null : val),
  
  device_group_id: z
    .number()
    .int()
    .positive()
    .nullable()
    .optional(),
});

/**
 * Device update form data type.
 */
export type DeviceUpdateFormData = z.infer<typeof deviceUpdateSchema>;

